src/pymor/algorithms/bfgs.py

```diff
@@ -0,0 +1,320 @@
+# License: BSD 2-Clause License (https://opensource.org/licenses/BSD-2-Clause)
+
+import warnings
+
+import numpy as np
+
+from pymor.algorithms.line_search import armijo
+from pymor.core.defaults import defaults
+from pymor.core.exceptions import BFGSError
+from pymor.core.logger import getLogger
+from pymor.parameters.base import Mu
+
+
+def get_active_and_inactive_sets(parameter_space, mu, epsilon=1e-8):
+    """Compute the active and inactive parameter index sets for constrained optimization.
+
+    Parameters
+    ----------
+    parameter_space
+        The |ParameterSpace| used for checking parameter ranges.
+    mu
+        The parameter to compute the active and inactive sets for.
+    epsilon
+        Tolerance threshold for checking boundary conditions.
+
+    Returns
+    -------
+    active
+        The set of active parameter indices.
+    inactive
+        The set of inactive parameter indices.
+    """
+    mu = mu if isinstance(mu, Mu) else parameter_space.parameters.parse(mu)
+    active = np.array([])
+
+    for key, dim in parameter_space.parameters.items():
+        indices = np.zeros(dim)
+        low_range = parameter_space.ranges[key][0] * np.ones(dim)
+        high_range = parameter_space.ranges[key][1] * np.ones(dim)
+        eps = epsilon * np.ones(dim)
+        indices[np.where(np.logical_or(mu[key] - low_range <= eps, high_range - mu[key] <= eps))] = 1
+        active = np.concatenate((active, indices))
+
+    inactive = np.ones(active.shape) - active
+
+    return active, inactive
+
+
+def compute_hessian_action_matrix(hessian, direction, active, inactive):
+    """Project the Hessian matrix applied to the direction using an active and an inactive set.
+
+    Parameters
+    ----------
+    hessian
+        The Hessian matrix.
+    direction
+        The descent direction used to update the parameter.
+    active
+        The active set computed from the |ParameterSpace|.
+    inactive
+        The inactive set computed from the |ParameterSpace|.
+
+    Returns
+    -------
+    hessian
+        The projection of the Hessian applied to the direction.
+    """
+    direction_A = np.multiply(active, direction)
+    direction_I = np.multiply(inactive, direction)
+    hessian = direction_A + np.multiply(inactive, hessian.dot(direction_I))
+    return hessian
+
+
+def update_hessian(hessian, mu, old_mu, gradient, old_gradient):
+    """Update Hessian matrix using BFGS iteration.
+
+    Parameters
+    ----------
+    hessian
+        The current Hessian matrix.
+    mu
+        The current `mu` parameter.
+    old_mu
+        The previous `mu` parameter.
+    gradient
+        The current gradient with respect to the parameter.
+    old_gradient
+        The previous gradient with respect to the parameter.
+
+    Returns
+    -------
+    hessian
+        The next BFGS Hessian matrix iterate.
+    """
+    gradient_difference = gradient - old_gradient
+    mu_difference = mu - old_mu
```

Comment:
So `mu` and `old_mu` apparently need to be a flattened `Mu` here. Should be mentioned in the docstring.